agent/consul/config.go

```diff
@@ -382,6 +394,60 @@ type Config struct {
 	ConnectReplicationToken string
 }
 
+// ResetSerfKey forcefully replaces all keys in the LAN and WAN keyrings with
+// the new key provided. It is only intended to be used for bootstrapping via
+// the /agent/bootstrap-gossip-key endpoint. It is here to save duplicating this
+// logic in both Client and Server implementations, and is called directly from
+// ReloadConfig in both cases.
+func (c *Config) ResetSerfKey(key []byte) error {
+	if c.SerfLANConfig.MemberlistConfig.Keyring == nil {
+		// Can't actually happen currently as we set an empty keyring even for
+		// disabled case but be defensive.
+		return errors.New("LAN gossip encryption not configured")
+	}
+	serfWANEnabled := false
+	// SerfWANConfig is actully non-nil on clients but has no keyring setup,
+	// however it can be explicitly nil on servers that have disabled WAN port.
+	if c.SerfWANConfig != nil && c.SerfWANConfig.MemberlistConfig != nil && c.SerfWANConfig.MemberlistConfig.Keyring != nil {
+		serfWANEnabled = true
+	}
+	err := c.SerfLANConfig.MemberlistConfig.Keyring.AddKey(key)
```

Comment:
Feels like you could add a local function that does the Add/Use/Get/Remove flow, and then call it once per Keyring to avoid such a long block of identical-ish code.

Reply:
Yeah but I wanted to ensure I can add and install on each without error before deleting etc. Seems marginally better to leave it in a state where LAN and WAN are different stages of a rotation rather than one completely rotated and using a totally different key not present in the other...

Either way it's not atomic so can get into unpleasant states and probably needs similar fix to be honest but it _seemed_ better this way 🤷‍♂️ .